include/ipc/ipc_service.h

```diff
@@ -157,21 +222,154 @@ int ipc_service_register_endpoint(const struct device *instance,
 
 /** @brief Send data using given IPC endpoint.
  *
- *  @param ept Registered endpoint by @ref ipc_service_register_endpoint.
- *  @param data Pointer to the buffer to send.
- *  @param len Number of bytes to send.
+ *  @param[in] ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param[in] data Pointer to the buffer to send.
+ *  @param[in] len Number of bytes to send.
  *
  *  @retval -EIO when no backend is registered or send hook is missing from
- *	    backend.
+ *               backend.
  *  @retval -EINVAL when instance or endpoint is invalid.
- *  @retval -EBADMSG when the message is invalid.
+ *  @retval -EBADMSG when the data is invalid (i.e. invalid data format,
+ *		     invalid length, ...)
  *  @retval -EBUSY when the instance is busy.
  *
  *  @retval bytes number of bytes sent.
  *  @retval other errno codes depending on the implementation of the backend.
  */
 int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len);
 
+/** @brief Get an empty TX buffer to be sent using @ref ipc_service_send_nocopy
+ *
+ *  This function can be called to get an empty TX buffer so that the
+ *  application can directly put its data into the sending buffer without copy
+ *  from an application buffer.
+ *
+ *  It is the application responsibility to correctly fill the allocated TX
+ *  buffer with data and passing correct parameters to @ref
+ *  ipc_service_send_nocopy function to perform data no-copy-send mechanism.
+ *
+ *  The size parameter can be used to request a buffer with a certain size. If
+ *  the size can be accomodated the function returns no errors and the size
+ *  value is unchanged. If the requested size is too big, the function returns
+ *  -ENOMEM and the size parameter must contain the maximum allowed size.
+ *
+ *  @param[in] ept Registered endpoint by @ref ipc_service_register_endpoint.
+ *  @param[out] data Pointer to the empty TX buffer.
+ *  @param[in,out] size Pointer to store the requested TX buffer size. If the
+ *			function returs -ENOMEM, this parameter returns the
+ *			maximum allowed size.
+ *  @param[in] wait Timeout waiting for an available TX buffer.
+ *
+ *  @retval -EIO when no backend is registered or send hook is missing from
+ *		 backend.
+ *  @retval -EINVAL when instance or endpoint is invalid.
+ *  @retval -ENOTSUP when this is not supported by backend.
+ *  @retval -ENOBUFS when there are no TX buffers available.
+ *  @retval -EALREADY when a buffer was already claimed and not yet released.
+ *  @retval -ENOMEM when the requested size is too big (and the size parameter
+ *		    contains the maximum allowed size).
+ *
+ *  @retval 0 on success.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_get_tx_buffer(struct ipc_ept *ept, void **data, uint32_t *size, k_timeout_t wait);
+
+/** @brief Drop and release a TX buffer
+ *
+ *  When @ref ipc_service_send_nocopy returns an error, it is possible to use
```

Comment:
Is it ok to use this function to drop a buffer without calling `ipc_serivce_send_nocopy` earlier?

I mean if a sequence like this is valid:

```
r = ipc_service_get_tx_buffer(...);

if (r == 0 && something_happend_causing_the_user_is_not_interested_in_sending_anymore()) {
  r = ipc_service_drop_tx_buffer(...);
}
```

From the description I understand it is not valid.

Reply:
I think the API should allow that. I'll clarify that use case.